### PR TITLE
Reduce the number of cache events in basic Reflex code; expose cheap …

### PR DESCRIPTION
**Pull request by danharaj**
…primitives in Reflex.Class; add a test for a GC bug that was discovered in the process of writing this patch

### CODE REVIEW

File: src/Reflex/Class.hs [hunk 1]
```diff
@@ -225,6 +234,8 @@ class ( MonadHold t (PushM t)
   -- 'Behavior's and hold 'Event's, and use the results to produce a occurring
   -- (Just) or non-occurring (Nothing) result
   push :: (a -> PushM t (Maybe b)) -> Event t a -> Event t b
+  -- | Like 'push' but intended for functions that the implementation can consider cheap to compute for performance considerations.
```
Comment (ryantrinkle) [inline src/Reflex/Class.hs]: Let's put a bigger, scarier comment on this.  We need to make it extremely clear that the user's function may be run multiple times.

File: src/Reflex/Patch.hs [hunk 2]
```diff
@@ -5,11 +5,14 @@
 {-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE StandaloneDeriving #-}
 {-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
 module Reflex.Patch
   ( Patch (..)
   , PatchDMap (..)
   , ComposeMaybe (..)
   , PatchMap (..)
+  , Group (..), Additive
```
Comment (ryantrinkle) [inline src/Reflex/Patch.hs]: Should `Additive` export `(..)`?  Also, should probably be on the next line.

File: src/Reflex/Patch.hs [hunk 3]
```diff
@@ -95,3 +86,18 @@ instance Ord k => Semigroup (PatchMap k v) where
 instance Ord k => Monoid (PatchMap k v) where
   mempty = PatchMap mempty
   mappend = (<>)
+
+-- Patches based on commutative groups
```
Comment (ryantrinkle) [inline src/Reflex/Patch.hs]: Haddock syntax?

File: src/Reflex/Pure.hs [hunk 4]
```diff
@@ -66,6 +66,8 @@ instance (Enum t, HasTrie t, Ord t) => Reflex (Pure t) where
   push :: (a -> PushM (Pure t) (Maybe b)) -> Event (Pure t) a -> Event (Pure t) b
   push f e = Event $ memo $ \t -> unEvent e t >>= \o -> f o t
 
+  pushCheap = push
```
Comment (ryantrinkle) [inline src/Reflex/Pure.hs]: Type signature.

File: src/Reflex/Spider/Internal.hs [hunk 5]
```diff
@@ -1641,9 +1666,12 @@ mergeCheap d = Event $ \sub -> do
...
           , subscriberRecalculateHeight = \_ -> return ()
           }
-    (EventSubscription _ changeSubd, change) <- subscribeAndRead (dynamicUpdated d) s
+    (EventSubscription sln changeSubscription, change) <- subscribeAndRead (dynamicUpdated d) s
     forM_ change $ \c -> defer $ SomeMergeUpdate (updateMe c) invalidateMyHeight revalidateMyHeight
-    liftIO $ writeIORef changeSubdRef (s, changeSubd)
+    -- We explicitly hold on to the unsubscribe function from subscribing to the update event.
+    -- If we don't do this, there are certain cases where mergeCheap will fail to properly retain
+    -- its subscription.
+    liftIO $ writeIORef changeSubdRef (s, sln, changeSubscription)
```
Comment (ryantrinkle) [inline src/Reflex/Spider/Internal.hs]: Let's not unpack and re-pack the EventSubscription, in line with the new comment above.

File: src/Reflex/Spider/Internal.hs [hunk 6]
```diff
@@ -1935,7 +1965,7 @@ instance HasSpiderTimeline x => R.Reflex (SpiderTimeline x) where
   {-# INLINABLE current #-}
   current = SpiderBehavior . dynamicCurrent . unSpiderDynamic
   {-# INLINABLE updated #-}
-  updated = SpiderEvent . fmap runIdentity . dynamicUpdated . unSpiderDynamic
+  updated = coerceWith Coercion $ SpiderEvent . dynamicUpdated . unSpiderDynamic
```
Comment (ryantrinkle) [inline src/Reflex/Spider/Internal.hs]: How is `coerceWith Coercion` different from `coerce`?

File: src/Reflex/Patch.hs (hunk 2 above)
Comment (danharaj) [inline src/Reflex/Patch.hs]: If you export with (..) for a class that doesn't have any members you get a warning. I can put it on another line tho.

Comment (ryantrinkle) [inline src/Reflex/Patch.hs]: Ah, ok; cool!

**Pull request merged**